app/routes/app/rewards/addresses.tsx

```diff
@@ -171,42 +159,47 @@ function AddressCards({ wallets }: { wallets: any }) {
   );
 }
 
-const schema = z.object({
-  payment: z.discriminatedUnion("tokenSymbol", [
-    z.object({ tokenSymbol: z.literal("ETH"), address: EthAddressSchema }),
-    z.object({ tokenSymbol: z.literal("SOL"), address: SolAddressSchema }),
-    z.object({ tokenSymbol: z.literal("MATIC"), address: EthAddressSchema }),
-  ]),
-});
+export const addWalletValidator = withZod(addWalletSchema);
+export const deleteWalletValidator = withZod(deleteWalletSchema);
+export const updateWalletValidator = withZod(updateWalletSchema);
 
-const validator = withZod(schema);
 function AddAddressButton() {
-  const { tokens } = useTypedLoaderData<typeof loader>();
+  const { networks } = useTypedLoaderData<typeof loader>();
   const [openedAdd, setOpenedAdd] = useState(false);
 
+  const isValid = useIsValid("addWallet");
+
   return (
     <>
       <Button className="mx-auto" onClick={() => setOpenedAdd(true)}>
         Add Address
       </Button>
       <Modal isOpen={openedAdd} onClose={() => setOpenedAdd(false)} title="Add an address">
-        <ValidatedForm validator={validator} className="space-y-5 mt-5">
+        <ValidatedForm
+          id="addWallet"
+          action="/app/rewards/wallet/add"
+          method="post"
+          validator={addWalletValidator}
+          className="space-y-5 mt-5"
+        >
           <div className="pb-44 pt-8">
-            <AddPaymentAddressForm tokens={tokens} />
+            <AddPaymentAddressForm networks={networks} />
           </div>
           <div className="flex gap-2 justify-end">
             <Button variant="cancel" onClick={() => setOpenedAdd(false)}>
               Cancel
             </Button>
-            <Button type="submit">Save</Button>
+            <Button disabled={!isValid} onClick={() => setOpenedAdd(false)} type="submit">
```

Comment:
I have a couple of ideas for a better UX here. First get rid of the disabled and onClick attributes.

```suggestion
            <Button type="submit">Save</Button>
```

This way validation works on submit press and doesn't immediately close the modal. Then we can add a default value to the ValidatedForm.

```ts
  defaultValues={{
    payment: {
      networkName: "Ethereum",
      address: "",
    },
  }}
  ```
  
 This does a couple of things 1) makes it so we don't need a placeholder text on the select 2) prevents the ugly "invalid discriminated union..." error from appearing (which I can't seem to find a way to override).

Reply:
Then we need to find a way to close the modal upon _successful_ wallet add. If there is a backend error (like the wallet already exist) then we have an opportunity to display the error in the modal.

To do that I think you'll need to bring the action inside the route file so that you can `useTypedActionData` to get a response.

Reply:
Good Idea, will play around with this now